Richard renaming netcdf_groundwater

RichardB9 opened the pull request:
This PR moves around and renames some modules in the `datasource` folder:

- `netcdf_groundwater.py` is renamed into `threedi_results.py`
- The class `NetcdfGroundwaterDataSource` is renamed into `ThreediResult`
- The `netcdf.py` is not used since the qgis 3 release and is now finally removed, some of its functions are moved into `threedi_results.py`
- Constants of threedi results are moved into `result_constants.py`

Commit:
Removed many unnecessary constants.

Commit:
Moved some methods within NetcdfGroundwaterDataSource class.

Commit:
Added extra class result_constants to store constants of result data

Commit:
Moved many constants into result_constants.py

Commit:
Removed VOLUME_AGG constant.

Commit:
Moved normalized_object_type, find_h5_file and find_aggregation_netcdf

Commit:
Moved detect_netcdf_version

Commit:
Forgotten from prev commit

Commit:
Added CUMULATIVE_AGGREGATION_UNITS which got lost during refactoring.

Commit:
Removed helper_functions_env.py.

Not used anywhere and seems like a local helper module for a dev.

Commit:
Removed netcdf.py!

Commit:
Renamed netcdf_groundwater to threedi_results and NetcdfGroundwaterDataSource to ResultData

Commit:
Renamed ResultData into ThreediResult

Commit:
Renamed netcdf_groundwater_ds fixture to threedi_result

Commit:
Removed unused method find_aggregation_netcdf

Commit:
Renamed find_aggregation_netcdf_gw to find_aggregation_netcdf

Commit:
Renamed some methods/vars still called netcdf_groundwater

Commit:
Fixed some imports which went wrong with previous refactors.

Commit:
Removed unused methods from ThreediResult

Commit:
Refactored available_aggregation_vars and made available_vars a property

Commit:
Merge branch 'master' of github.com:nens/ThreeDiToolbox into richard_datasource_cleanup

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+from collections import namedtuple
+
+# NetCDF variable information
+NcVar = namedtuple("NcVar", ["name", "verbose_name", "unit"])
```
Ik heb zelf namedtuples nog nooit gebruikt ("ik maak wel een class of een dict"), maar op deze manier zien ze er toch wel èrg makkelijk uit.
Hoog tijd om ze zelf ook te gaan gebruiken :-)

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+    SOURCES_AND_SINKS,
+]
+
+AGGREGATION_OPTIONS = {
```
Waarom is dit een set? Een list is vaak makkelijker/leesbaarder (in django hebben ze de tuples in de settings ook door lijsten vervangen, bijvoorbeeld).

RichardB9 replied:
Omdat dubbele items niet voorkomen en de volgorde van de items niet van belang is.

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+}
+
+# Explanation: aggregation using the cumulative method integrates the variable
+# over time. Therefore the units must be multiplied by the time.
```
```suggestion
# over time. Therefore the units must be multiplied by time.
```

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+
+layer_information = [
+    # layer name, (normalized) object_type, q/h type
+    # Note: the reason why this is plural is because this is (inconsistently)
```
Met "this" bedoel je denk ik het eerste item (connection_nodes) en niet de hele lijst?
Dan is dit duidelijker:

```suggestion
    # Note: the reason why v2_connection_nodes is plural is because this is (inconsistently)
```

RichardB9 replied:
Dat weet ik eigenlijk niet zeker, ik heb het zo overgenomen. Ik zie helemaal onderaan ook andere layer names in meervoud (bijv. 'pumplines').
Denk wel dat de comment outdated is omdat een 'id mapping json' niet meer gebruikt wordt.

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+    ("node_results", "node", "h"),
+]
+
+layer_object_type_mapping = dict([(a[0], a[1]) for a in layer_information])
```
Hoofdletters? (Zelfde voor volgende regel).

reinout commented on datasource/result_constants.py:
```diff
@@ -0,0 +1,129 @@
+}
+
+layer_information = [
+    # layer name, (normalized) object_type, q/h type
```
Is een namedtuple hier niet ook lekker? Dan zit de uitleg in de definitie v/d namedtuple i.p.v. in een commentaarregel.

RichardB9 replied:
Ja goed idee!

reinout commented on datasource/threedi_results.py:
```diff
@@ -7,74 +7,19 @@
 
+from .result_constants import layer_object_type_mapping
 from .base import BaseDataSource
-from ..utils import cached_property
```
De wijzigingen in deze imports gaan even bijten met mijn imports-veranderd-pull-request.
Even opletten aan beide kanten.

reinout commented on datasource/threedi_results.py:
```diff
@@ -7,74 +7,19 @@
-    find_h5_file,
-)
-from ..utils.user_messages import messagebar_message
+from .result_constants import SUBGRID_MAP_VARIABLES
```
TODO erbij zetten dat we nog naar de naam moeten kijken.

reinout commented on datasource/threedi_results.py:
```diff
@@ -99,47 +44,20 @@ def __init__(self, file_path=None):
-                self.gridadmin_result.nodes._field_names
-                | self.gridadmin_result.lines._field_names
-                | self.gridadmin_result.pumps._field_names
+                    self.gridadmin_result.nodes._field_names
```
Dit zal niet aan deze PR letten, maar..... underscore variabelen zijn normaliter expliciet verboden terrein voor externen.
En nu zitten we in underscore variabelen van een variabele van een variabele te "wroeten"...

Is dit het goede gebruik van de gridadmin?

reinout commented on datasource/threedi_results.py:
```diff
@@ -148,37 +66,28 @@ def available_subgrid_map_vars(self):
         """Return a list of available variables in the 'aggregate_results_3di.nc"""
-        agg = self.gridadmin_aggregate_result
-        if not agg:
+        ga = self.gridadmin_aggregate_result
```
Is `ga` hier een goed idee? **DE** gridadmin bestaat toch niet? We hebben er twee. Om nou dezelfde term voor verschillende dingen te gebruiken lijkt me niet lekker.

En als het het resultaat is (i.p.v. de gridadmin zelf, tenminste, zo ziet het er qua naamgeving uit) kan je het beter `aggregate_result` ofzo noemen?

reinout commented on datasource/threedi_results.py:
```diff
@@ -219,8 +128,29 @@ def get_timestamps(self, parameter=None):
+        if variable is None:
+            return self.gridadmin
+        elif variable in self.available_subgrid_map_vars:
+            return self.gridadmin_result
```
"gridadmin_result" is "het resultaat van de gridadmin".
"result_gridadmin" maakt het duidelijker dat het een gridadmin is?
Ik weet niet, ik zit nog met de namen te spelen.

`regular_gridadmin` en `aggregate_gridadmin`?

Of is het meer "resultaat" dan "gridadmin"? 

`regular_result` en `aggregate_result`?

RichardB9 replied:
Als we kijken naar `threedigrid` worden de volgende klassen gebruikt:
1. `GridH5Admin`
2. `GridH5ResultAdmin`
3. `GridH5AggregateResultAdmin`

De volgende namen zijn dan misschien wel consequent:
1. `grid_h5_admin`
2. `grid_h5_result_admin`
3. `grid_h5_aggregate_result_admin`

Maar de namen worden toch wel erg lang, iets korter:
1. `gridadmin` (alleen `admin` kan misleidend zijn)
2. `result_admin`
3. `aggregate_result_admin`

reinout commented on datasource/threedi_results.py:
```diff
@@ -404,12 +314,23 @@ def gridadmin_result(self):
             return None
 
+    @property
+    def ds(self):
```
Misschien al eerder gevraagd, maar kan dit `def datasource()` worden i.p.v. `ds`?

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
+def find_h5_file(netcdf_file_path):
+    """An ad-hoc way to get the h5_file.
+
+    We assume the h5_file file is in on of the following locations (note:
```
```suggestion
    We assume the h5_file file is in one of the following locations (note:
```

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
+        h5_file path
+
+    Raises:
+        IndexError if nothing is found
```
Is een `FileNotFound` error niet veel explicieter?

IndexError is een zijeffect van de gekozen methode van zoeken, niet een duidelijke error.

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
         try:
-            aggregation_netcdf_file = find_aggregation_netcdf_gw(self.file_path)
+            aggregation_netcdf_file = find_aggregation_netcdf(self.file_path)
         except IndexError:
```
Zie commentaar iets verderop over IndexError: dat zou een FileNotFound error moeten zijn.
Hier staat de indexerror zo raar/onduidelijk.

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
+    pattern = "*.h5"
+    inpdir = os.path.join(os.path.dirname(netcdf_file_path), "..", "preprocessed")
+    resultdir = os.path.dirname(netcdf_file_path)
+
```
Zoiets lijkt me netter:

    for directory in [inpdir, resultdir]:
        h5_files = glob.glob(..._
        if h5_files:
            return h5files[0]
    raise FileNotFound(zoiets, even kijken welke error het precies is)

Eventueel een eigen kleine Exception voor aanmaken. H5NotFound. Kan ook.

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
+    """
+    pattern = "aggregate_results_3di.nc"
+    result_dir = os.path.dirname(netcdf_file_path)
+    return glob.glob(os.path.join(result_dir, pattern))[0]
```
Ook hier weer een impliciete IndexError.

reinout commented on datasource/threedi_results.py:
```diff
@@ -420,10 +341,101 @@ def ds_aggregation(self):
+            - 'netcdf-groundwater'
+
+    """
+    import h5py
```
Waarom h5py op dit punt importeren?

RichardB9 replied:
Ah, kan weg :)

reinout commented on test/test_datasources.py:
```diff
@@ -35,7 +34,7 @@
 )
 
-netcdf_groundwater_datasource_nc_path = os.path.join(
+results_3di_path = os.path.join(
```
`threedi_results_path` is taalkundig netter. Volgorde.

En: hoofdletters van maken?

reinout commented on test/test_datasources.py:
```diff
@@ -128,24 +111,24 @@ def test_import_layer(self):
 
     def test_sanity(self):
-        nds = NetcdfGroundwaterDataSource()
+        nds = ThreediResult()
```
nds is de afko v/d oude class.

Ik zou er gewoon (in de hele file) `threedi_result` van maken.

reinout commented on test/test_datasources.py:
```diff
@@ -154,157 +137,193 @@ def test_find_agg_fail(self):
-    np.testing.assert_equal(ts[:, 0], netcdf_groundwater_ds.get_timestamps())
-    assert ts.shape[1] == netcdf_groundwater_ds.get_gridadmin("s1").nodes.count + 1
+def test_get_timeseries(threedi_result):
+    ts = threedi_result.get_timeseries("s1")
```
`timeseries`? Daar zit namelijk in dat het ook wel eens meerdere timeseries kunnen zijn.
`ts` kan als enkelvoud `timeserie` gelezen worden en daardoor voor foutjes zorgen.

reinout commented on test/test_datasources.py:
```diff
@@ -154,157 +137,193 @@ def test_find_agg_fail(self):
+def test_available_vars(threedi_result):
+    actual = threedi_result.available_vars
+    normal_vars = {
+        'au', 'q', 'q_lat', 'q_pump', 'rain', 's1', 'su', 'u1', 'vol',
```
"Black" zal er nog even overheen moeten: enkele quotes :-)

reinout commented on threedi_statistics/tools/statistics.py:
```diff
@@ -71,7 +71,7 @@ def nFlowLine(self):
     @property
     def has_groundwater(self):
-        return self.obj.__class__.__name__ == "NetcdfGroundwaterDataSource"
+        return self.obj.__class__.__name__ == "ThreediResult"
```
```suggestion
        return isinstance(self.obj, ThreediResult)
```

Commit:
Renaming some things in result_constants.py.

Commit:
Added todo to look at a constant name.

Commit:
Renamed threedi_result.ds to datasource

Commit:
Updated find_h5_file to raise FileNotFoundError.

Commit:
Update find_aggregation_netcdf to raise FileNotFoundError

Commit:
Removed h5py import in method.

Commit:
Renamed some constants/variables

Commit:
Renamed ts to time_series and update has_groundwater check.

Commit:
Ran Black over test-file.

Commit:
Renamed:
- gridadmin_result to result_admin
- gridadmin_aggregate_result to aggregate_result_admin

Commit:
Merge branch 'master' of github.com:nens/ThreeDiToolbox into richard_datasource_cleanup

Commit:
Fix some imports which went wrong during merge.
